src/api/useTodo.tsx

```diff
@@ -0,0 +1,82 @@
+import React, { useMemo, useState } from 'react';
+import { Error } from '../types/Error';
+import { Filter } from '../types/Filter';
+import { Todo } from '../types/Todo';
+
+interface TodosGlobalContext {
+  todos: Todo[];
+  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>;
+  filteredTodos: Todo[];
+  errorMessage: Error;
+  setErrorMessage: (e: Error) => void;
+  filter: Filter;
+  setFilter: (f: Filter) => void;
+  tempTodo: Todo | null;
+  setTempTodo: (t: Todo | null) => void;
+  toBeCleared: number[];
+  setCleared: (d: number[]) => void;
+}
+
+export const TodosContext = React.createContext<TodosGlobalContext>({
+  todos: [],
+  setTodos: () => {},
+  filteredTodos: [],
+  errorMessage: Error.Absent,
+  setErrorMessage: () => {},
+  filter: Filter.all,
+  setFilter: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  toBeCleared: [],
+  setCleared: () => {},
+});
+
+export const useTodo = (): TodosGlobalContext => React.useContext(TodosContext);
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<Error>(Error.Absent);
+  const [filter, setFilter] = useState<Filter>(Filter.all);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [toBeCleared, setCleared] = useState<number[]>([]);
+
+  const filteredTodos = useMemo(() => {
+    return todos.filter(todo => {
+      switch (filter) {
+        case Filter.active:
+          return !todo.completed;
+
+        case Filter.completed:
+          return todo.completed;
+
+        case Filter.all:
+        default:
+          return true;
+      }
+    });
+  }, [todos, filter]);
```

Comment:
No need to filter in case Filter is All. make filter() call and return in `case`, by default return `todos`